plugins/kubernetes-backend/src/service/KubernetesProxy.ts

```diff
@@ -0,0 +1,286 @@
+        return response;
+      }),
+    );
+
+    const data: { [key: string]: any } = {};
+    const codes: number[] = [];
+
+    responses.forEach(kpr => {
+      if (kpr.cluster) {
+        data[kpr.cluster] = kpr.data;
+        codes.push(kpr.code);
+      }
+    });
+
+    const code = this.getBestResponseCode(codes);
+
+    const res: KubernetesProxyResponse = {
+      code,
+      data,
+    };
+
+    return res;
+  }
+
+  protected getKubernetesRequestedClusters(
+    req: Request,
+  ): KubernetesProxyClusters {
+    const encodedClusters: string =
+      req.header(HEADER_KUBERNETES_CLUSTERS) ?? '';
+
+    if (!encodedClusters) {
+      return {};
+    }
+
+    try {
+      const decodedClusters = Buffer.from(encodedClusters, 'base64').toString();
+      const clusters: KubernetesProxyClusters = JSON.parse(decodedClusters);
+      return clusters;
+    } catch (e: any) {
+      this.logger.error(
+        `error with encoded cluster header: ${JSON.stringify(e)}`,
+      );
+    }
+    return {};
+  }
+
+  protected async getClusterDetails(
+    clusterSupplier: KubernetesClustersSupplier,
+    krc: KubernetesProxyClusters,
+  ): Promise<ClusterDetails[]> {
+    const clusters = await clusterSupplier.getClusters();
+
+    const clusterNames = Object.keys(krc);
+
+    const clusterDetails = clusters.filter(c => clusterNames.includes(c.name));
+
+    const clusterDetailsAuth = await Promise.all(
+      clusterDetails.map(async c => {
+        const cAuth: ClusterDetails = Object.assign(c, {
+          serviceAccountToken: krc[c.name],
+        });
+        return cAuth;
+      }),
+    );
+
+    return clusterDetailsAuth;
+  }
+
+  protected getClusterURI(details: ClusterDetails): string {
+    const client = this.getKubeConfig(details);
+    return client.getCurrentCluster()?.server || '';
+  }
+
+  protected async makeRequestToCluster(
+    details: ClusterDetails,
+    req: Request,
+  ): Promise<KubernetesProxyResponse> {
+    const serverIP = this.getClusterURI(details);
+    if (!serverIP) {
+      this.logger.error(
+        `could not find server IP for cluster with name ${details.name}`,
+      );
+      return {
+        code: ERROR_INTERNAL_SERVER,
+        data: null,
+      };
+    }
+
+    const query = decodeURIComponent(req.params.encodedQuery) || '';
+    const uri = `${serverIP}/${query}`;
+
+    const contentType = req.header(HEADER_CONTENT_TYPE) || APPLICATION_JSON;
+
+    const res = await this.sendClusterRequest(
+      details,
+      uri,
+      req.method,
+      contentType,
+      req.body,
+    );
+
+    return res;
+  }
+
+  protected async sendClusterRequest(
+    details: ClusterDetails,
+    uri: string,
+    method: string,
+    contentType: string,
+    body?: any,
+  ): Promise<KubernetesProxyResponse> {
+    const bearerToken = details.serviceAccountToken;
+    if (!bearerToken) {
+      this.logger.error('could not find bearer token!');
```

Comment:
Probably not useful to output logs for these cases, right? The operator reading these logs after the fact - at error level - won't find them actionable or having enough context / information. At best this could be at debug level to aid in local testing use cases. But in general, most logs in this file should probably be removed